### PR TITLE
[NPUW] Support avoids for no partitioning case

### PR DESCRIPTION
**Pull request by smirnov-alexey**

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp [hunk 1]
```diff
@@ -286,6 +289,9 @@ void Snapshot::fuseRemnants() {
             for (const auto& cons : consumers) {  // FIXME: pick the smallest flops
                 Group::GPtr cons_group = m_graph->meta(cons).get<Group::GPtr>();
                 if (!group->hasCycle(cons_group)) {
+                    if (group->avoidedTargets() != cons_group->avoidedTargets()) {
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp:292]: Might be better to add to a condition below..

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 2]
```diff
@@ -410,7 +415,11 @@ class Compiler {
 
         switch (currentPipeline()) {
         case Pipeline::NONE:
-            break;  // unreachable
+            // Allow partitioning to merge everything into a single group
+            ctx.min_graph_size = 1;
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:419]: To be honest, I expect this `ctx.min_graph_size` set in `none()` call

Comment (AsyaPronina) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:419]: With `ctx` in `setCtx()`

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 3]
```diff
@@ -323,7 +323,17 @@ class Compiler {
...
+        // Note: If there are avoids present there will be multiple groups formed instead of 1
+        m_snapshot->earlyAvoids();
+        m_snapshot->repeat([&] {
+            m_snapshot->repeat([&] {
+                m_snapshot->collectLHF();
+            });
+            m_snapshot->repeat([&] {
+                m_snapshot->fuseRemnantsExtended();
+            });
+        });
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:336]: I'd check if there's any `AVOID`s, then go hard way, otherwise `singleGroup()` could still work.

`AVOID`s are still an exception rather than somewhat normal

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 4]
```diff
@@ -387,11 +397,6 @@ class Compiler {
         : m_model(model),
           m_snapshot(std::make_shared<Snapshot>(model)),
           m_cfg(cfg) {
-        if (currentPipeline() == Pipeline::NONE) {
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:390]: Can check for avoids here and proceed with single model pipeline

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp [hunk 5]
```diff
@@ -339,6 +345,10 @@ void Snapshot::fuseInputs() {
             }
             // Found 2 inputs to fuse
             if (inputs_to_fuse.first && inputs_to_fuse.second) {
+                if (inputs_to_fuse.first->avoidedTargets() != inputs_to_fuse.second->avoidedTargets()) {
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp:348]: Add test on specific pattern

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 2 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:419]: `none()`, `rep()`, etc only set pass logic and order. `ctx` manipulations are done exactly here for all pipelines (isolates, avoids, nofold, etc). So I would like to keep this logic

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 3 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 4 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:390]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp (hunk 5 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/snapshot.cpp:345]: Done

**Comment by smirnov-alexey** [comment]
Tested locally PIPELINE:NONE w/ and w/o the avoids. The results are as expected. Ready to be reviewed/merged

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 6]
```diff
@@ -328,6 +328,25 @@ class Compiler {
         LOG_INFO("Done");
     }
 
+    void none_with_avoids() {
+        LOG_INFO("Online partitioning: compiling single group with avoids pipeline...");
+        LOG_BLOCK();
+
+        // Note: Assuming the context was set with minimum graph size = 1
```
Comment (AsyaPronina) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:154]: Still bad assumption, I am sorry. Seems like responsibility for `none` pipeline is split between code and if something changed, both places need to be updated. I think this code can be named not `none`, but with some other name. Or context can be changed here. At least we can add assert, but I don't recommend that way.

**Review by AsyaPronina — COMMENTED** [review]
Please either add functional tests or create a JIRA for coming Sprint to add tests.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 6 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp:154]: We don't pass the context to those pipeline-defining function and I'd prefer to keep it this way. The responsibility of setting the context falls onto the Compiler constructor while those function should only define the pass logic

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 7]
```diff
@@ -328,6 +328,25 @@ class Compiler {
...
+        // Note: If there are avoids present there will be multiple groups formed instead of 1
+        m_snapshot->earlyAvoids();
+        m_snapshot->repeat([&] {
+            m_snapshot->repeat([&] {
+                m_snapshot->collectLHF();
+            });
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: I don't think we even need this here.. why?
Can't a repeated `fuseRemnantsExtended()` do the job?

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 8]
```diff
@@ -387,7 +406,15 @@ class Compiler {
...
           m_snapshot(std::make_shared<Snapshot>(model)),
           m_cfg(cfg) {
-        if (currentPipeline() == Pipeline::NONE) {
+        PassContext ctx;
+        ctx.min_graph_size = detail::getMinGraphSize(m_cfg);
+        ctx.keep_blocks = detail::getMinRepBlocks(m_cfg);
+        ctx.keep_block_size = detail::getMinRepBlockSize(m_cfg);
+        ctx.avoids = detail::getAvoids(m_cfg);
+        ctx.isolates = detail::getIsolates(m_cfg);
+        ctx.nofolds = detail::getNoFolds(m_cfg);
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: Probably it's time to
```suggestion
        PassContext ctx(m_cfg);
```

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 9]
```diff
@@ -387,7 +406,15 @@ class Compiler {
...
+
+        if (currentPipeline() == Pipeline::NONE && ctx.avoids.empty()) {
             none();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: Probably it'd be right to rename this one to `none_fast()`

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp [hunk 10]
```diff
@@ -398,19 +425,15 @@ class Compiler {
...
-
         m_snapshot->setCtx(ctx);
 
         switch (currentPipeline()) {
         case Pipeline::NONE:
-            break;  // unreachable
+            // Allow partitioning to merge everything into a single group
+            ctx.min_graph_size = 1;
+            m_snapshot->setCtx(ctx);
+            none_with_avoids();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: and keep this one as `none()`

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/group.cpp [hunk 11]
```diff
@@ -272,10 +267,6 @@ void Group::fuseInputs(const std::pair<Group::GPtr, Group::GPtr>& gptr_inputs) {
     Group::GPtr absorbing_group = gptr_inputs.first;
     Group::GPtr absorbed_group = gptr_inputs.second;
 
-    if (absorbing_group->m_avoided_devices != absorbed_group->m_avoided_devices) {
-        return;
-    }
-
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/online/group.cpp:278]: What's the reason to these changes? Probably I asked already

File: src/plugins/intel_npu/tests/unit/npuw/online_partitioning.cpp [hunk 12]
```diff
@@ -525,3 +525,48 @@ TEST(OnlinePartitioningTest, Partitioning_Compiler_Compute_RepeatedModel) {
...
+    snap->setCtx(ctx);
+    snap->earlyAvoids();
+    snap->repeat([&] {
+        snap->repeat([&] {
+            snap->collectLHF();
+        });
+        snap->repeat([&] {
+            snap->fuseRemnantsExtended();
+        });
+    });
```
Comment (dmatveev) [inline src/plugins/intel_npu/tests/unit/npuw/online_partitioning.cpp]: abstraction leak here, why it is not the `Compiler` with `Pipeline::NONE`?

Comment (smirnov-alexey) [reply src/plugins/intel_npu/tests/unit/npuw/online_partitioning.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/group.cpp (hunk 11 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/group.cpp:278]: I've moved them to the passes which use `fuse` and `fuseWith`. As we discussed there was a corner case which prevented the correct fusing. The exact example is in the test cases.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 10 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 9 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 8 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp (hunk 7 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/online/compiler.cpp]: Done
